lib/symboldatabase.cpp

```diff
@@ -4202,8 +4204,41 @@ static const Token* skipPointersAndQualifiers(const Token* tok)
     return tok;
 }
 
+const Token* skipTypeOrFunctionIdentifier(const Token* tok)
+{
+    if (tok == nullptr)
+        return nullptr;
+    while (tok->next() != nullptr && (tok->next()->str() == "::" || tok->next()->str() == "<"))
+    {
+        if (tok->next()->str() == "::")
+            tok = tok->tokAt(2);
+        else
+            tok = tok->linkAt(1);
+        if (tok == nullptr)
+            return nullptr;
+    }
+    return tok->next();
+}
+
+//Could this token be the first token of a call to a function instead of the declaration of a variable
+bool isFunctionCallInsteadOfVariableDeclaration(const Token* tok, const Scope& scope, const SymbolDatabase& db)
+{
+    auto start = tok;
+    tok = skipTypeOrFunctionIdentifier(tok);
+    if (tok == nullptr)
+        return false; //It's probably neither but it's not a function call
+    if (tok->str() != "(") //X y isn't a function call
+        return false;
+    if (tok->astOperand2() != nullptr && tok->astOperand2()->str() == ",")// X(a, b) isn't a variable declaration
+        return true;
+    return db.findFunction(start, &scope) != nullptr;
+}
+
 bool Scope::isVariableDeclaration(const Token* const tok, const Token*& vartok, const Token*& typetok) const
 {
+    if (isFunctionCallInsteadOfVariableDeclaration(tok, *this, *this->check))
+        return false;
```

Comment:
This is the crux of the change

Reply:
Can you clarify what you mean? I'm not sure how to action this. this method is in Scope not SymbolDatabase but I suspect that isn't really the point.

Reply:
It would be preferable if the `Tokenizer::setVarId` could distinguish when there is a function call or if there is a variable declaration. However in `Tokenizer::setVarId` we do not have type info so it is more difficult. Maybe some "simple" logic can be added there; if there is a function with the name "foo" then do not guess that "foo" is a type. It seems weird/unlikely to reuse the same name for a type and a function.

Reply:
action: could you at least take a look at `Tokenizer::setVarId` and determine if it seems possible to improve it.

Reply:
I'm not sure varId is the issue. Compare these two snippets:
```
int* a = new int;
void foo(int x){}
int main() {
    foo(*a); //This should be interpreted as a function call and is by cppcheck
}
```
```
int* a[1] = { new int };
void foo(int x){}
int main() {
    foo(*a[0]); //This should be interpreted as a function call but is not by cppcheck
}
```

For a bunch of weird permutations take a look at https://godbolt.org/z/5zhWrYfs9

The first case will pass the test because it correctly identifies foo on the 4th line as a funtion. The second case will incorrectly fail to identify foo on the 4th line as a function. However in both cases mVarId==0 for the token in question. To me this suggests that setVarId is doing the right thing or, at least, isn't causing this problem.

Reply:
Actually I think I might be at least partially wrong about this. It loos like in the second example the second ```a```'s mVarId is initially 3 and then gets erroneously cleared to zero during createSymbolDatabaseSetVariablePointers. This makes cppcheck think it isn't a variable reference anymore. So mVarId may or may not be _the_ problem but it's certainly _a_ problem.

Reply:
specifically addressing the question of whether the code needs to exist in the symboldatabase I'd like to highlight lines 14 and 17 in https://godbolt.org/z/5zhWrYfs9. Syntactically these are identical. The only thing that distinguishes them is the existence of a function called ```bar```. With this in mind, you can't decide whether such a line is a function call or variable declaration until you know for sure whether a function with that name already exists. I think, but am not sure, that this means  you need to do it when setting up the symboldatabase.

Reply:
@danmar thoughts?